### ISSUE TITLE
refactor: chainrules cleanup — remove Variable API, unify scalar AD rules, add oracle tests

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#534]
## Summary

Comprehensive cleanup of the `chainrules` / `chainrules-scalarops` crates and their integration with tenferro's scalar AD layer. Two parallel AD APIs exist, scalar AD rules are duplicated, and oracle-based test coverage is missing.

## Background

Discussion identified:
1. **Two parallel AD APIs**: Tape-first (`Tape<V>` + `TrackedValue<V>`) vs PyTorch-like (`Variable<V>` + `AutogradGraph<V>`). Only Tape-first is used by tenferro's production facade.
2. **DRY violation**: `tenferro/src/ops/scalar/ad/unary.rs` reimplements AD math (sqrt, exp, log, sin, cos, etc.) at tensor level using `AnalyticUnaryOp` dispatch + `mul_with_conj_factor`, duplicating what `chainrules-scalarops` provides at scalar level.
3. **Missing rules**: `chainrules-scalarops` only has frule/rrule for conj, sqrt, add, sub, mul, div, powf, powi. Missing: exp, expm1, log, log1p, sin, cos, tanh, asin, acos, atan, sinh, cosh, asinh, acosh, atanh.
4. **No oracle-based tests**: [tensor-ad-oracles](https://github.com/tensor4all/tensor-ad-oracles) has reference data (finite-difference + PyTorch) for all these ops but is not used in chainrules-scalarops tests.

---

### What to delete

The `Variable<V>` / `AutogradGraph<V>` API is a PyTorch-style monomorphic AD graph that is **not used by tenferro's production facade**. Only `Tape<V>` + `TrackedValue<V>` is used.

#### Files to delete entirely

| File | Lines | Content |
|------|-------|---------|
| `extern/chainrules/src/engine/variable.rs` | ~302 | `Variable<V>` struct, `backward()`, `zero_grad()`, `detach()`, `requires_grad_()` etc. |
| `extern/chainrules/src/ops/autograd_common.rs` | 57 | `merge_context_for_binary_op`, `merge_context_for_multi_op` — only used by `autograd.rs` Variable ops |
| `extern/chainrules/src/ops/test_support.rs` | 15 | `square_graph()` — builds `Variable`-based test graph |
| `extern/chainrules/tests/autodiff_next_api_tests.rs` | 73 | Tests `Variable`, `AutogradGraph` API |
| `extern/chainrules/tests/autodiff_next_accum_tests.rs` | 88 | Tests Variable gradient accumulation |
| `extern/chainrules/tests/autodiff_next_grad_query_tests.rs` | 57 | Tests Variable gradient query API |
| `extern/chainrules/tests/autodiff_next_lifetime_tests.rs` | 46 | Tests Variable lifetime semantics |
| `extern/chainrules/tests/autodiff_next_contract_errors.rs` | 32 | Tests Variable contract errors |
| `tenferro-einsum/tests/variable_einsum_tests.rs` | 188 | Tests `variable_einsum` |

#### Files to modify

| File | Change |
|------|--------|
| `extern/chainrules/src/ops/autograd.rs` (452 lines) | **Delete**: `AddRule` (L11-57), `SquareRule` (L59-90), `add()` (L157-205), `square()` (L218-269), `grad_tangent()` (L277-347), `grad_variable()` (L361-452). **Keep only**: `record_op()` (L108-136) — used by `tenferro-einsum/src/ad/tracked.rs:189`. Move `record_op` to a simpler location or inline. |
| `extern/chainrules/src/lib.rs` | Remove from re-exports: `Variable`, `AutogradGraph`, `BackwardOptions`. Remove `pub use ops::test_support;`. Keep `pub use ops::autograd;` only if `record_op` stays there. |
| `extern/chainrules/src/engine/mod.rs` | Remove `variable` module declaration and re-export. Remove `VariableNodeKind` if Variable-only. |
| `extern/chainrules/src/engine/context.rs` | Check if `AutogradGraph` is Variable-only. If so, delete. If shared with Tape, keep. |
| `tenferro-einsum/src/ad/tracked.rs` | Remove `variable_einsum` function (L131-190). Keep `tracked_einsum`. Update import of `autograd::record_op`. |
| `tenferro-einsum/src/ad/mod.rs` | Remove `variable_einsum` re-export. |
| `tenferro-einsum/src/lib.rs` | Remove `variable_einsum` re-export. |

#### Verification

The `record_op` function in `autograd.rs` is the **only production code** in the Variable module. It's used at exactly one callsite:
```
tenferro-einsum/src/ad/tracked.rs:189:    autograd::record_op(output, operands, Box::new(rule), tangent_out)
```
This function takes `&[&Variable<V>]` as input — it will need to be refactored or the `Variable` type kept minimally for this signature. Check whether `tracked.rs:189` actually uses `Variable` or can be migrated to `TrackedValue`.

---

### What exists

`extern/chainrules-scalarops/src/unary.rs` (190 lines) already has frule/rrule for `conj`, `sqrt`, `exp`, `log`. But `exp`/`log` rules are **untested**.

`extern/chainrules-scalarops/src/binary.rs` (205 lines) has frule/rrule for `add`, `sub`, `mul`, `div`.

### What to add

Add frule/rrule to `extern/chainrules-scalarops/src/unary.rs` for:

| Function | frule formula | rrule formula |
|----------|--------------|---------------|
| expm1 | `df = exp(x) * dx` | `grad = ḡ * exp(x)` |
| log1p | `df = dx / (1+x)` | `grad = ḡ / (1+x)` |
| sin | `df = cos(x) * dx` | `grad = ḡ * cos(x)` |
| cos | `df = -sin(x) * dx` | `grad = -ḡ * sin(x)` |
| tanh | `df = (1 - tanh²(x)) * dx` | `grad = ḡ * (1 - tanh²(x))` |
| asin | `df = dx / √(1-x²)` | `grad = ḡ / √(1-x²)` |
| acos | `df = -dx / √(1-x²)` | `grad = -ḡ / √(1-x²)` |
| atan | `df = dx / (1+x²)` | `grad = ḡ / (1+x²)` |
| sinh | `df = cosh(x) * dx` | `grad = ḡ * cosh(x)` |
| cosh | `df = sinh(x) * dx` | `grad = ḡ * sinh(x)` |
| asinh | `df = dx / √(x²+1)` | `grad = ḡ / √(x²+1)` |
| acosh | `df = dx / √(x²-1)` | `grad = ḡ / √(x²-1)` |
| atanh | `df = dx / (1-x²)` | `grad = ḡ / (1-x²)` |

### Complex convention (Wirtinger)

For complex types, the pattern follows what's already in `chainrules-scalarops`: the derivative factor gets `.conj()` applied (Wirtinger ∂f/∂z̄). See existing `sqrt_frule`/`sqrt_rrule` for the pattern:
```rust
// frule: tangent = dx * conj(factor)
// rrule: grad = g * conj(factor)
```
where `factor = df/dz` computed from the primal.

### Signature convention

Follow the existing pattern in `unary.rs`:
```rust
pub fn sin_frule<T: ScalarAd>(x: T, dx: T) -> (T, T)   // returns (primal, tangent)
pub fn sin_rrule<T: ScalarAd>(x: T, g: T) -> T          // returns grad_x
```
Note: some rrules (like `sqrt_rrule`) take the primal output `y` instead of input `x` when that's more efficient. Choose whichever avoids redundant computation.

---

### Current architecture (to be replaced)

`tenferro/src/ops/scalar/ad/unary.rs` (225 lines) uses a macro `define_conj_factor_unary_builder!` that:
1. Computes primal via `AnalyticUnaryOp` dispatch (goes through runtime context → CPU backend)
2. Computes a "factor" tensor using tensor-level operations (e.g., for sin: `cos(input)`)
3. Multiplies tangent/cotangent by factor with `mul_with_conj_factor`

This is **16 builders** (SqrtAdBuilder, ExpAdBuilder, ..., AtanhAdBuilder), each reimplementing the derivative formula using tensor primitive calls.

### Target architecture

Replace with element-wise application of `chainrules_scalarops` frule/rrule. The pattern is **already established** in `tenferro/src/core/dynamic/dyn_ad_tensor/scalar_ops.rs`:

```rust
// Existing pattern for mul (scalar_ops.rs L221-240):
fn scale_ad_tensor_typed<T>(tensor: &AdTensor<T>, scalar: &AdTensor<T>) -> Result<AdTensor<T>>
where T: Scalar + ScalarAd + Copy + DynTensorTyped + 'static
{
    let (primal, tangent) = tensor_binary_scalar_ad_typed(
        tensor.primal(),
        tensor.tangent(),
        extract_rank0_scalar(scalar, "scale")?,
        extract_rank0_scalar_tangent(scalar, "scale")?,
        chainrules_scalarops::mul,       // primal fn
        chainrules_scalarops::mul_frule, // frule fn
    )?;
    merge_tensor_scalar_output(tensor, scalar, primal, tangent,
        chainrules_scalarops::mul_rrule, // rrule fn
    )
}
```

For unary ops (sin, exp, etc.), the equivalent pattern using `tensor_map_unary_typed`:
```rust
// Conceptual replacement for SinAdBuilder:
fn sin_ad_typed<T>(input: &AdTensor<T>) -> Result<AdTensor<T>>
where T: Scalar + ScalarAd + Copy + DynTensorTyped + 'static
{
    let primal = tensor_map_unary_typed(input.primal(), chainrules_scalarops::sin)?;
    let tangent = input.tangent().map(|dt| {
        tensor_map_binary_typed(input.primal(), dt, |x, dx| {
            let (_, dy) = chainrules_scalarops::sin_frule(x, dx);
            dy
        })
    }).transpose()?;
    // For reverse: register rrule using chainrules_scalarops::sin_rrule per-element
    ...
}
```

### Verification

All 16 unary ops have existing finite-difference tests in `tenferro/src/ops/ad/tests/scalar_generic.rs` (358 lines). These tests must continue to pass after the replacement.

---

### Data source

[tensor-ad-oracles](https://github.com/tensor4all/tensor-ad-oracles) provides JSONL oracle data with this structure per case:

```json
{
  "case_id": "sin_f64_identity_001",
  "dtype": "float64",
  "inputs": { "a": { "shape": [20, 20], "data": [...] } },
  "probes": [{
    "direction":   { "value": { "data": [...], "shape": [20,20] } },
    "cotangent":   { "value": { "data": [...], "shape": [20,20] } },
    "fd_ref":      { "jvp": { "data": [...] }, "vjp": { "data": [...] } },
    "pytorch_ref": { "jvp": { "data": [...] }, "vjp": { "data": [...] } }
  }],
  "comparison": {
    "first_order":  { "atol": 1e-08, "rtol": 1e-05, "kind": "allclose" },
    "second_order": { "atol": 1e-09, "rtol": 0.0001, "kind": "allclose" }
  }
}
```

### Integration approach (TBD — decide one)

1. **Git submodule**: Add `tensor-ad-oracles` as a submodule under `extern/` or `tests/fixtures/`
2. **Build-script fetch**: Download oracle JSONL at test time via HTTP
3. **Vendored fixtures**: Copy relevant JSONL files into `extern/chainrules-scalarops/tests/fixtures/`

Option 1 (submodule) is cleanest for keeping data in sync. Option 3 is simplest but risks staleness.

### Test harness

Add to `extern/chainrules-scalarops/tests/`:
```rust
// For each op (sin, cos, exp, ...):
// 1. Load JSONL from oracle cases/<op>/identity.jsonl
// 2. Parse input tensor, direction, cotangent, fd_ref, pytorch_ref
// 3. For each element:
//    - Verify: op_frule(input[i], direction[i]).1 ≈ pytorch_ref.jvp[i]  (atol/rtol from comparison)
//    - Verify: op_rrule(input[i], cotangent[i]) ≈ pytorch_ref.vjp[i]
// 4. Also cross-check against fd_ref for extra confidence
```

Oracle cases to use (matching chainrules-scalarops functions):

| chainrules-scalarops fn | Oracle case path |
|------------------------|------------------|
| sqrt | `cases/sqrt/identity.jsonl` |
| exp | `cases/exp/identity.jsonl` |
| expm1 | `cases/expm1/identity.jsonl` |
| log | `cases/log/identity.jsonl` |
| log1p | `cases/log1p/identity.jsonl` |
| sin | `cases/sin/identity.jsonl` |
| cos | `cases/cos/identity.jsonl` |
| tanh | `cases/tanh/identity.jsonl` |
| asin | `cases/asin/identity.jsonl` |
| acos | `cases/acos/identity.jsonl` |
| atan | `cases/atan/identity.jsonl` |
| sinh | `cases/sinh/identity.jsonl` |
| cosh | `cases/cosh/identity.jsonl` |
| asinh | `cases/asinh/identity.jsonl` |
| acosh | `cases/acosh/identity.jsonl` |
| atanh | `cases/atanh/identity.jsonl` |
| add | `cases/add/identity.jsonl` |
| sub | `cases/sub/identity.jsonl` |
| mul | `cases/mul/identity.jsonl` |
| div | `cases/div_no_rounding_mode/identity.jsonl` |
| conj | `cases/conj/identity.jsonl` |

---

## Phase ordering

```
Phase 1 (independent)
  └──→ Phase 2a (independent of Phase 1)
         └──→ Phase 2b (requires 2a: needs the new frule/rrule functions)
         └──→ Phase 2c (requires 2a: tests the new frule/rrule functions)
```

Phase 1 and Phase 2a can be done in parallel. Phase 2b and 2c both depend on 2a.

## Related

- tensor4all/tensor-ad-oracles#16 — add analytic formula documentation (after Phase 2a)
- PR #532 (merged) — dead code cleanup
- PR #533 (merged) — scalar type list unification
